proxier: only get local addresses once per sync loop

andrewsykim opened the pull request:
Signed-off-by: Andrew Sy Kim <kiman@vmware.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Currently the iptables and userspace proxier lists all local network interfaces every time it checks if an external IP is local to the system. For most clusters this isn't problematic but for clusters using a lot of external IPs this can impact kube-proxy performance significantly. An optimization can be made where we get all local addresses only once per sync and store it in memory.

The IPVS proxy made this optimization in https://github.com/kubernetes/kubernetes/pull/84924, this PR makes the same optimization for the iptables proxy. More details in https://github.com/kubernetes/kubernetes/issues/84906

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/issues/84906

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
iptables/userspace proxy: improve performance by getting local addresses only once per sync loop, instead of for every external IP
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

andrewsykim commented:
/sig network
/assign @danwinship

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -992,9 +997,9 @@ func (proxier *Proxier) syncProxyRules() {
-				klog.Errorf("can't determine if IP is local, assuming not: %v", err)
-			} else if local && (svcInfo.Protocol() != v1.ProtocolSCTP) {
+			if len(localAddrs) == 0 {
+				klog.Errorf("couldn't find any local IPs, assuming external IP %s is not local", externalIP)
```
You already logged an error about `GetLocalAddrs` failing above. Do we need to log an error for each external IP too? (Maybe just add "assuming all external IPs are not local" to the earlier message.)
(Likewise in ipvs.)

andrewsykim replied:
Good catch, thanks

danwinship reviewed (commented):
Maybe fix the userspace proxy code too and then you can get rid of `utilproxy.IsLocalIP`?

andrewsykim commented:
I was hesitate to do this for userspace proxy because it wasn't scoped to only external IPs, will update it though

andrewsykim committed:
iptables proxier: get local addresses only once per sync loop

This avoids fetching all local network interfaces everytime we sync an
external IP. For clusters with many external IPs this gets really
expensive. This change caches all local addresses once per sync.

Signed-off-by: Andrew Sy Kim <kiman@vmware.com>

andrewsykim committed:
ipvs proxier: use util proxy methods for getting local addresses

Signed-off-by: Andrew Sy Kim <kiman@vmware.com>

andrewsykim committed:
userspace proxy: get local addresses only once per sync loop

This avoids fetching all local network interfaces everytime we sync an
external IP. For clusters with many external IPs this gets really
expensive. This change caches all local addresses once per sync.

Signed-off-by: Andrew Sy Kim <kiman@vmware.com>

andrewsykim committed:
proxier: use IPSet from k8s.io/utils/net to store local addresses

This allows the proxier to cache local addresses instead of fetching all
local addresses every time in IsLocalIP.

Signed-off-by: Andrew Sy Kim <kiman@vmware.com>

andrewsykim commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -903,10 +908,7 @@ func (proxier *Proxier) closePortal(service proxy.ServicePortName, info *Service
 	el := []error{}
-
-	if local, err := utilproxy.IsLocalIP(portal.ip.String()); err != nil {
-		el = append(el, fmt.Errorf("can't determine if IP %s is local, assuming not: %v", portal.ip, err))
```
This error is no longer being added to `el`, I think this is fine but not familiar enough with userspace proxy to be sure.

andrewsykim commented:
@danwinship comments addressed, PTAL :)

andrewsykim commented:
/assign @thockin

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -758,6 +758,11 @@ func (proxier *Proxier) syncProxyRules() {
+	localAddrs, err := utilproxy.GetLocalAddrs()
+	if err != nil {
+		klog.Errorf("Failed to get local addresses during proxy sync: %v, assuming external IPs are not local", err)
+	}
```
Should we log if len(localAddrs) == 0 ?  You specifically tested for it below, but I think that would be a nice thing to log.

andrewsykim replied:
done

thockin commented on pkg/proxy/util/utils.go:
```diff
@@ -123,23 +123,36 @@ func IsProxyableHostname(ctx context.Context, resolv Resolver, hostname string)
-// on the local system
-func IsLocalIP(ip string) (bool, error) {
+// GetLocalAddrs returns a list of all network addresses on the local system
+func GetLocalAddrs() ([]net.IP, error) {
```
Why not return a "set" type here.  k8s.io/utils/net/ipnet.go defines `IPNetSet`.  We could have the equivalent `IPSet` type

andrewsykim replied:
Makes sense to me, I'll open a PR to add `IPSet` type to k8s.io/utils/net.

andrewsykim replied:
On second thought, I'm realizing that a set type for net.IP may not work because we can't do comparison checks for net.IP based on their string values since an IPv6 representation of IPv4 addresses are considered equal (see [net.IP Equals](https://golang.org/src/net/ip.go?s=10028:10057#L397)). We could just ignore this case and assume equality/uniqueness soley based on string values but that would be a behavioral change from what we have right now.

danwinship replied:
But `String` has the same logic. Any two `net.IP`s that compare as `Equal` will also `String`-ify the same. (Specifically, `net.ParseIP("::ffff:1.2.3.4").String() → "1.2.3.4"`)

So as long as you know the IPs are in canonical form, you can compare based on string values.

andrewsykim replied:
Ah good to know! I'll update k8s.io/utils/net to add a set type for net.IP.

thockin commented on pkg/proxy/util/utils.go:
```diff
@@ -123,23 +123,36 @@ func IsProxyableHostname(ctx context.Context, resolv Resolver, hostname string)
+}
+
+// ContainsIP returns true if ip exists in addrs, false otherwise
+func ContainsIP(ip net.IP, ips []net.IP) bool {
```
This is why I think a set type makes sense.

Also, the "receiver" arg should go first if we don't have a type.

thockin reviewed (commented):
relatively minor points.

andrewsykim committed:
update vendor k8s.io/utils to 5f6fbceb4c31

Signed-off-by: andrewsykim <kim.andrewsy@gmail.com>

andrewsykim commented:
Blocked on https://github.com/kubernetes/kubernetes/issues/87408

andrewsykim commented:
Ready for another review

andrewsykim commented:
/priority important-soon

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1030,9 +1040,7 @@ func (proxier *Proxier) syncProxyRules() {
-			if local, err := utilproxy.IsLocalIP(externalIP); err != nil {
-				klog.Errorf("can't determine if IP is local, assuming not: %v", err)
-			} else if local && (svcInfo.Protocol() != v1.ProtocolSCTP) {
+			if localAddrSet.Len() > 0 && (svcInfo.Protocol() != v1.ProtocolSCTP) && localAddrSet.Has(net.ParseIP(externalIP)) {
```
nit: Do we really need to check `.Len() == 0` ?

andrewsykim replied:
No this check was here before the set type was added so I think it's redundant now. I'll remove this in a follow-up PR

thockin reviewed (commented):
Thanks!

/lgtm
/approve
